pageserver/src/tenant.rs

```diff
@@ -1063,34 +1344,71 @@ impl Tenant {
         .with_context(|| {
             format!(
                 "Failed to create timeline {}/{} metadata",
-                new_timeline_id, self.tenant_id
+                self.tenant_id, new_timeline_id
             )
         })?;
 
-        let ancestor = new_metadata
-            .ancestor_timeline()
-            .and_then(|ancestor_timeline_id| timelines.get(&ancestor_timeline_id))
-            .cloned();
         let new_timeline = self
-            .initialize_new_timeline(new_timeline_id, new_metadata, ancestor)
+            .create_timeline_data(new_timeline_id, new_metadata, ancestor)
```

Comment:
If error appears here, then uninit mark is dropped, and timeline will be considered healthy during restart? Am I correct?

Reply:
> I would imagine that timeline is completely initialized in temporary location (with all initdb loading) and then moved to permatent location with a rename.

That is a great idea and I've tried doing it, but the issue is that we have to write into the tmp directory during initdb result imports. 
That would require is to pass the knowledge about the tmp dir from `Tenant` level to `DeltaLayer` (and others') level: right now, we import the data and do checkpointing. During checkpointing, we use `conf.rs` functions to get the timeline path and derive the layer path to write into out of it. 
Changing that would require a lot of code, unless I'm mistaken somehow.

So, that's why the whole tombstone approach as a "simpler" way of doing things.

Reply:
> If error appears here, then uninit mark is dropped, and timeline will be considered healthy during restart? Am I correct?

I don't think so: with the first step we create the timeline dir, then we would check that in the mark drop logic and keep the mark.
But you're right that we can make it better and at least try to remove the files right away, if fail to create the timeline: if we succeed, we could try to create this timeline before the next restart.

I've moved the `after-timeline-uninit-mark-creation` failpoint to this code and adjusted the corresponding test to not to restart pagerver at all.

Reply:
> then we would check that in the mark drop logic and keep the mark.

You're right. I've missed it. To me tombstone approach is way harder to reason about. Indeed many changes are required to pass around different base directory. Not sure exactly how many, I havent tried it. Hard to say is it worth it to make more changes but make it easier (subjectively) to reason about it

Reply:
Also coverage thinks that this branch in drop was not covered:
https://neondatabase.github.io/zenith-coverage-data/02d93d71a0ee9b43225ad4dc15613106ff62c4c2/coverage/pageserver/src/tenant.rs.html#L315

Is it coverage problem or test doesnt trigger this case?

Reply:
Neither, it's a special handle: current PR logic actually tries to remove the timeline dir first before removing the mark, and only if the timeline dir removal fails the mark code that you've linked would be triggered.
Not sure how realistic this is, but decided it's safer to keep the mark aware of this.

Reply:
Timelines' init tmp dir approach issue: https://github.com/neondatabase/neon/issues/2670